resource-thief.c

```diff
@@ -0,0 +1,361 @@
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/types.h>
+#include <signal.h>
+#include <fcntl.h>
+
+#include <gbm.h>
+#include <pool-buffer.h>
+#include <wayland-client-protocol.h>
+#include <linux-dmabuf-unstable-v1-client-protocol.h>
+
+#define POOL_SIZE 1024
+#define POINTS_PER_REGION 300
+#define DMABUF_WIDTH 64
+#define DMABUF_HEIGHT 64
+#define DMABUF_FORMAT GBM_FORMAT_XRGB8888
+
+static void handle_sigint(int sig) {
+	(void)sig;
+}
+
+/* Different types of resources to trick the compositor into exhausting */
+enum mode {
+	CONSUME_NOOP,
+	CONSUME_MAPS, /* wl_shm_pool */
+	CONSUME_MEMORY, /* wl_region */
+	CONSUME_FDS /* linux_dmabuf? */
+};
+
+struct connection {
+	struct wl_list link;
+	uint32_t resource_count;
+	struct wl_display *display;
+	struct wl_registry *registry;
+	struct wl_shm *shm;
+	struct wl_compositor *compositor;
+	struct zwp_linux_dmabuf_v1 *linux_dmabuf;
+	struct wl_shm_pool **pools;
+	struct wl_region **regions;
+	struct wl_buffer **buffers;
+};
+
+void destroy_connection(struct connection *conn) {
+	if (!conn) {
+		return;
+	}
+	if (conn->pools) {
+		for (uint32_t i = 0; i < conn->resource_count; i++) {
+			if (conn->pools[i]) {
+				wl_shm_pool_destroy(conn->pools[i]);
+			}
+		}
+	}
+	if (conn->buffers) {
+		for (uint32_t i = 0; i < conn->resource_count; i++) {
+			if (conn->buffers[i]) {
+				wl_buffer_destroy(conn->buffers[i]);
+			}
+		}
+	}
+	if (conn->regions) {
+		for (uint32_t i = 0; i < (conn->resource_count / POINTS_PER_REGION + 1); i++) {
+			if (conn->regions[i]) {
+				wl_region_destroy(conn->regions[i]);
+			}
+		}
+	}
+	if (conn->shm) {
+		wl_shm_destroy(conn->shm);
+	}
+	if (conn->compositor) {
+		wl_compositor_destroy(conn->compositor);
+	}
+	if (conn->linux_dmabuf) {
+		zwp_linux_dmabuf_v1_destroy(conn->linux_dmabuf);
+	}
+	if (conn->registry) {
+		wl_registry_destroy(conn->registry);
+	}
+	if (conn->display) {
+		wl_display_disconnect(conn->display);
+	}
+	free(conn->pools);
+	free(conn);
+}
+
+
+static void reg_global(void *data, struct wl_registry *registry,
+		uint32_t num, const char *str, uint32_t version) {
+	struct connection *conn = (struct connection *)data;
+	if (!strcmp(str, "wl_shm")) {
```

Comment:
Please use `strcmp(…) == 0` to make the comparison more readable. Use `wl_shm_interface.name` instead of a plain string to make it easier to avoid typos.